### PR TITLE
driver: stm32: usb: add usb detach

### LINKED ISSUE #51474: driver: stm32: usb: add detach function support
**Is your enhancement proposal related to a problem? Please describe.**
STM32 USB driver does not implement USB detach function

**Describe the solution you'd like**
Implement it using STM32 HAL

### PR DESCRIPTION
**Pull request by RomainPelletant**
USB detach was not implemented in stm32 usb driver.
Add support, tested on STM32L5

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/51474

### CODE REVIEW

**Comment by desowin** [comment]
@Light411 Could you please check if it solves your DFU issue?

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -992,6 +992,17 @@ int usb_dc_detach(void)
...
+	status = HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
+	if (status != HAL_OK) {
+		LOG_ERR("PCD_DeInit failed, %d", (int)status);
+		return -EIO;
+	}
+
+	if (irq_is_enabled(USB_IRQ)) {
+		irq_disable(USB_IRQ);
+	}
+
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c:1021]: Looks good, but don't we need to undo everything done in attach(), like disabling clocks enabled by usb_dc_stm32_clock_enable().

**Comment by Light411** [comment]
Could we compare the approach taken in this PR with the one proposed in #36848 ? 
Is that better to use HAL_PCD_DevDisconnect or HAL_PCD_DeInit ? 

I can try that this afternoon. For now my quick fix was using #36848 implementation.

**Comment by erwango** [comment]
> Could we compare the approach taken in this PR with the one proposed in #36848 ? Is that better to use HAL_PCD_DevDisconnect or HAL_PCD_DeInit ?
> 
> I can try that this afternoon. For now my quick fix was using #36848 implementation.

According to SUB API, `usb_dc_detach` is linked to `usb_disable` which expactations are minimal:
```
 * @brief Disable the USB device
 *
 * Function to disable the USB device.
 * Upon success, the specified USB interface is clock gated in hardware,
 * it is no longer capable of generating interrupts.
```
So in theory `HAL_PCD_DevDisconnect` should be sufficient, I guess.

Though, since `usb_dc_attach` will perform a whole init sequence including setting up the clocks, `HAL_PCD_Init`, `HAL_PCD_Stop/Start`, calling `HAL_PCD_DeInit` in `usb_dc_detach` should be safe.

Maybe the new USB API will be an occasion to perform actions with finer grain.

**Comment by RomainPelletant** [comment]
@loicpoulain 
Shall I call only the following clock function or implement other related PLL CLK activations (USB_CLK_SRC, RCC_MSI, etc...)?
```
if (clock_control_off(clk, (clock_control_subsys_t *)&pclken) != 0) {
     LOG_ERR("Unable to enable USB clock");
     return -EIO;
}
```

@erwango 
Do you think I shall abort that PR?

**Comment by erwango** [comment]
> @loicpoulain Shall I call only the following clock function or implement other related PLL CLK activations (USB_CLK_SRC, RCC_MSI, etc...)?
> 
> ```
> if (clock_control_off(clk, (clock_control_subsys_t *)&pclken) != 0) {
>      LOG_ERR("Unable to enable USB clock");
>      return -EIO;
> }
> ```

All code around related domain clocks should be revisited to be configured by device tree and centralized using clock_control framework. For instance turning off MSI, if not done through a centralized procedure could have consequences on other consumers of this clock.
I advise not touching this. `clock_control_off` should be sufficient for now. 

> @erwango Do you think I shall abort that PR?

No, please continue it.

**Comment by RomainPelletant** [comment]
PR has been updated with requested modifications

**Comment by desowin** [comment]
@RomainPelletant When you update your commit to latest main you should rebase it on top of main instead of merging the main branch into your branch.

**Comment by RomainPelletant** [comment]
@desowin thanks for the advice. Rebase triggered using github button

**Review by erwango — DISMISSED** [review]

**Review by desowin — DISMISSED** [review]

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Commit message should be:
```
drivers: usb_dc_stm32: implement usb_dc_detach()

Implement support to detach USB device.
```

**Review by jfischer-no — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -348,6 +348,22 @@ static int usb_dc_stm32_clock_enable(void)
...
+static int usb_dc_stm32_clock_disable(void)
+{
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	struct stm32_pclken pclken = {
+		.bus = USB_CLOCK_BUS,
+		.enr = USB_CLOCK_BITS,
+	};
+
+	if (clock_control_off(clk, (clock_control_subsys_t *)&pclken) != 0) {
+		LOG_ERR("Unable to enable USB clock");
```
Comment (FRASTM) [inline drivers/usb/device/usb_dc_stm32.c:360]: LOG_ERR("Unable to **disable** USB clock");

**Review by FRASTM — CHANGES_REQUESTED** [review]
else LGTM

**Review by FRASTM — DISMISSED** [review]

**Review by desowin — DISMISSED** [review]

**Review by jfischer-no — DISMISSED** [review]

**Comment by erwango** [comment]
Rebasing to fix CI issues

**Review by jfischer-no — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Review by desowin — APPROVED** [review]

**Pull request merged**